pydantic_ai_slim/pydantic_ai/_parts_manager.py

```diff
@@ -152,6 +170,127 @@ def handle_text_delta(
+
+                # Handle any text before the start tag
+                text_event = None
+                if before_start:
+                    text_event = self._handle_text_delta_simple(
+                        vendor_part_id=vendor_part_id,
+                        content=before_start,
+                        id=id,
+                        ignore_leading_whitespace=ignore_leading_whitespace,
+                    )
+
+                # Clear any state for this vendor_part_id and start thinking part
+                self._vendor_id_to_part_index.pop(vendor_part_id, None)
+                self._tag_buffer.pop(vendor_part_id, None)
+                thinking_event = self.handle_thinking_delta(vendor_part_id=vendor_part_id, content='')
+
+                # Process any remaining content after the start tag recursively
+                if after_start:
+                    self._handle_text_delta_with_thinking_tags(
+                        vendor_part_id=vendor_part_id,
+                        content=after_start,
+                        id=id,
+                        thinking_tags=thinking_tags,
+                        ignore_leading_whitespace=ignore_leading_whitespace,
+                    )
+                    # Return the first event that was created (text part or thinking part)
+                    return text_event if text_event is not None else thinking_event
+                else:
+                    # No content after start tag
+                    return text_event if text_event is not None else thinking_event
+            elif self._could_be_tag_start(combined_content, start_tag):
+                # Might be start of start tag, buffer it
+                if vendor_part_id is not None:
+                    self._tag_buffer[vendor_part_id] = combined_content
+                return None
+            else:
+                # Not a start tag, process as normal text
+                if vendor_part_id is not None:
+                    self._tag_buffer.pop(vendor_part_id, None)
+                return self._handle_text_delta_simple(
+                    vendor_part_id=vendor_part_id,
+                    content=combined_content,
+                    id=id,
+                    ignore_leading_whitespace=ignore_leading_whitespace,
+                )
+
+    def _could_be_tag_start(self, content: str, tag: str) -> bool:
+        """Check if content could be the beginning of a tag.
+
+        This is used to determine whether we should buffer content or process it immediately.
+        We check if the tag starts with the content, which means the content could be
+        a partial tag that will be completed in a future chunk.
+        """
+        if not content:
+            return False
+        # Check if the tag starts with any suffix of the content
+        # E.g., for content="<thi" and tag="<think>", we check if "<think>" starts with "<thi"
+        for i in range(len(content)):
```

Comment:
We don't need to look at the entire content, right, just the last `len(tag)` chars?

Reply:
This was replaced by the function `_parts_manager.py::_could_be_tag_start()`

```python
    def _could_be_tag_start(self, content: str, tag: str) -> bool:
        """Check if content could be the start of a tag."""
        # Defensive check for content that's already complete or longer than tag
        # This occurs when buffered content + new chunk exceeds tag length
        # Example: buffer='<think' + new='<' = '<think<' (7 chars) >= '<think>' (7 chars)
        if len(content) >= len(tag):
            return False
        return tag.startswith(content)
```